x-pack/plugins/canvas/public/socket.js

```diff
@@ -10,14 +10,55 @@ import { loadBrowserPlugins } from './lib/load_browser_plugins';
 
 let socket;
 
-export function createSocket(basePath) {
-  socket = io(undefined, { path: `${basePath}/socket.io` });
+export async function createSocket(basePath) {
+  if (socket != null) return socket;
+
+  const status = {};
+
+  socket = io({
+    path: `${basePath}/socket.io`,
+    transports: ['polling', 'websocket'],
+    transportOptions: {
+      polling: {
+        extraHeaders: {
+          'kbn-xsrf': 'professionally-crafted-string-of-text',
+        },
+      },
+    },
+    rememberUpgrade: false,
+    timeout: 5000,
+  });
 
   socket.on('getFunctionList', () => {
     const pluginsLoaded = loadBrowserPlugins();
 
     pluginsLoaded.then(() => socket.emit('functionList', functionsRegistry.toJS()));
   });
+
+  socket.on('connect', () => {
+    status.resolve();
+  });
+
+  function errorHandler(err) {
+    if (!err instanceof Error) {
```

Comment:
Quirky bit of logic... I get it, but it took a second.  Consider revising for clarity.

Reply:
I wrote it backwards originally, but it's the exceptional case. I can assign the comparison to a variable to make it clearer though.